contracts/global-extensions/GlobalOptimisticAuctionRebalanceExtension.sol

```diff
@@ -0,0 +1,328 @@
+     * @param _newComponentsAuctionParams   AuctionExecutionParams for new components, indexed corresponding to _newComponents.
+     * @param _oldComponentsAuctionParams   AuctionExecutionParams for existing components, indexed corresponding to
+     *                                      the current component positions. Set to 0 for components being removed.
+     * @param _shouldLockSetToken           Indicates if the rebalance should lock the SetToken.
+     * @param _rebalanceDuration            Duration of the rebalance in seconds.
+     * @param _positionMultiplier           Position multiplier at the time target units were calculated.
+    */
+    function proposeRebalance(
+        ISetToken _setToken,
+        IERC20 _quoteAsset,
+        address[] memory _oldComponents,
+        address[] memory _newComponents,
+        AuctionExecutionParams[] memory _newComponentsAuctionParams,
+        AuctionExecutionParams[] memory _oldComponentsAuctionParams,
+        bool _shouldLockSetToken,
+        uint256 _rebalanceDuration,
+        uint256 _positionMultiplier
+    )
+        external
+    {
+        bytes32 proposalHash = keccak256(abi.encode(
+            _setToken,
+            _quoteAsset,
+            _oldComponents,
+            _newComponents,
+            _newComponentsAuctionParams,
+            _oldComponentsAuctionParams,
+            _shouldLockSetToken,
+            _rebalanceDuration,
+            _positionMultiplier
+        ));
+        ProductSettings memory settings = productSettings[_setToken];
+
+        require(assertionIds[proposalHash] == bytes32(0), "Proposal already exists");
+        require(settings.rulesHash != bytes32(""), "Rules not set");
+        require(address(settings.optimisticParams.optimisticOracleV3) != address(0), "Oracle not set");
+
+
+        bytes memory claim = _constructClaim(proposalHash, settings.rulesHash);
+
+  
+
+        
+        uint256 minimumBond = settings.optimisticParams.optimisticOracleV3.getMinimumBond(address(settings.optimisticParams.collateral));
+        uint256 totalBond =  minimumBond > settings.optimisticParams.bondAmount ? minimumBond : settings.optimisticParams.bondAmount;
+
+        settings.optimisticParams.collateral.safeTransferFrom(msg.sender, address(this), totalBond);
+        settings.optimisticParams.collateral.safeIncreaseAllowance(address(settings.optimisticParams.optimisticOracleV3), totalBond);
+
+        bytes32 assertionId = settings.optimisticParams.optimisticOracleV3.assertTruth(
+            claim,
+            msg.sender,
+            address(this),
+            address(0),
+            settings.optimisticParams.liveness,
+            settings.optimisticParams.collateral,
+            settings.optimisticParams.bondAmount,
+            settings.optimisticParams.identifier,
+            bytes32(0)
+        );
+
+        assertionIds[proposalHash] = assertionId;
+        proposalHashes[assertionId] = proposalHash;
+
```

Comment:
event emission?

Reply:
I agree with @pblivin0x some extra event emissions would probably be nice. Probably each of the external state changing methods could use a "success" event emission at the end.

Reply:
Included now.